audioswitch/src/main/java/com/twilio/audioswitch/AudioDeviceSelector.kt

```diff
@@ -1,79 +1,107 @@
+    private val EARPIECE_AUDIO_DEVICE = AudioDevice(EARPIECE, "Earpiece")
+    private val SPEAKERPHONE_AUDIO_DEVICE = AudioDevice(SPEAKERPHONE, "Speakerphone")
+    private val WIRED_HEADSET_AUDIO_DEVICE = AudioDevice(WIRED_HEADSET, "Wired Headset")
+    private var bluetoothAudioDevice: AudioDevice? = null
+    internal var state: State = STOPPED
+    internal enum class State {
         STARTED, ACTIVATED, STOPPED
     }
-    private val EARPIECE_AUDIO_DEVICE = AudioDevice(AudioDevice.Type.EARPIECE, "Earpiece")
-    private val SPEAKERPHONE_AUDIO_DEVICE = AudioDevice(AudioDevice.Type.SPEAKERPHONE, "Speakerphone")
-    private val WIRED_HEADSET_AUDIO_DEVICE = AudioDevice(AudioDevice.Type.WIRED_HEADSET, "Wired Headset")
-    private var bluetoothAudioDevice: AudioDevice? = null
-    private val bluetoothController: BluetoothController? = BluetoothController.newInstance(
-            context,
-            LogWrapper(),
-        object : BluetoothDeviceConnectionListener {
-            override fun onBluetoothConnected(
-                bluetoothDevice: BluetoothDevice
-            ) {
-                bluetoothAudioDevice = AudioDevice(
-                        AudioDevice.Type.BLUETOOTH,
-                        bluetoothDevice.name)
-                if (state == State.ACTIVATED) {
-                    userSelectedDevice = bluetoothAudioDevice
-                }
-                enumerateDevices()
+    internal val bluetoothDeviceConnectionListener = object : BluetoothDeviceConnectionListener {
+        override fun onBluetoothConnected(
+            bluetoothDevice: BluetoothDevice
+        ) {
+            bluetoothAudioDevice = AudioDevice(
+                    BLUETOOTH,
+                    bluetoothDevice.name)
+            if (state == ACTIVATED) {
+                userSelectedDevice = bluetoothAudioDevice
             }
+            enumerateDevices()
+        }
 
-            override fun onBluetoothDisconnected() {
-                bluetoothAudioDevice = null
-                enumerateDevices()
+        override fun onBluetoothDisconnected() {
+            bluetoothAudioDevice = null
+            enumerateDevices()
+        }
+    }
+    internal val wiredDeviceConnectionListener = object : WiredDeviceConnectionListener {
+        override fun onDeviceConnected() {
+            wiredHeadsetAvailable = true
+            logger.d(TAG, "Wired Headset available")
+            if (this@AudioDeviceSelector.state == ACTIVATED) {
+                userSelectedDevice = WIRED_HEADSET_AUDIO_DEVICE
             }
+            enumerateDevices()
+        }
+
+        override fun onDeviceDisconnected() {
+            wiredHeadsetAvailable = false
+            enumerateDevices()
+        }
+    }
+
+    companion object {
+
+        /**
+         * Constructs a new AudioDeviceSelector instance.
+         *
+         * @param context the application context
+         */
+        fun newInstance(context: Context): AudioDeviceSelector {
```

Comment:
What is the reason for the static creation method versus a constructor?

Reply:
So I initially tried implementing two constructors, one internal and one public. It resulted in code that looked something like this:

```
    internal constructor(
            logger: LogWrapper,
            audioDeviceManager: AudioDeviceManager,
            wiredHeadsetReceiver: WiredHeadsetReceiver,
            bluetoothController: BluetoothController?) {
        this.logger = logger
        this.audioDeviceManager = audioDeviceManager
        this.wiredHeadsetReceiver = wiredHeadsetReceiver
        this.bluetoothController = bluetoothController
    }

    constructor(context: Context) {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        val logger = LogWrapper()
        val audioDeviceManager =
                AudioDeviceManager(context,
                        logger,
                        audioManager,
                        BuildWrapper(),
                        AudioFocusRequestWrapper())
        this.logger = logger
        this.audioDeviceManager = audioDeviceManager
        this.wiredHeadsetReceiver = WiredHeadsetReceiver(context, logger)
        this.bluetoothController = BluetoothController.newInstance(context, logger, audioDeviceManager)
    }

    private val logger: LogWrapper
    private val audioDeviceManager: AudioDeviceManager
    private val wiredHeadsetReceiver: WiredHeadsetReceiver
    private val bluetoothController: BluetoothController?
```

Unfortunately you have to duplicate the property setter calls and it becomes more verbose. When you use the static factory it allows you to invoke the class's primary constructor and you define the declaration and assignment of the properties all in one place, versus the method above which assigns them in two different constructors along with having to explicitly declare the properties outside of the constructors.

Reply:
And the reason we need two constructors is to allow our tests to inject their own collaborators to the class under test so we can assert/verify their behavior.